### PR TITLE
Restrictions on refresh

### LINKED ISSUE #448: Rate Limiting after 20 refreshes
Getting rate limited after running the script for a little more than 2 minutes every single time. tried pinging only a single district with multiple refresh rates ranging from 5 to 15, with the same result. did they make some changes in the backend?

### PR DESCRIPTION
**Pull request by 6mxo89sqh**
fixes #448 

Not tested. can someone test?

### CODE REVIEW

File: src/utils.py [hunk 1]
```diff
@@ -9,8 +9,8 @@
 
 BOOKING_URL = "https://cdn-api.co-vin.in/api/v2/appointment/schedule"
 BENEFICIARIES_URL = "https://cdn-api.co-vin.in/api/v2/appointment/beneficiaries"
-CALENDAR_URL_DISTRICT = "https://cdn-api.co-vin.in/api/v2/appointment/sessions/calendarByDistrict?district_id={0}&date={1}"
-CALENDAR_URL_PINCODE = "https://cdn-api.co-vin.in/api/v2/appointment/sessions/calendarByPin?pincode={0}&date={1}"
+CALENDAR_URL_DISTRICT = "https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/calendarByDistrict?district_id={0}&date={1}"
+CALENDAR_URL_PINCODE = "https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/calendarByPin?pincode={0}&date={1}"
```
Comment (shailesh) [inline src/utils.py:13]: Why are we updating Public APIs here since we know they have 30 mins of cache?

Comment (Nakul93) [inline src/utils.py:13]: If we are using public API, won't there be any issue of caching?

Comment (6mxo89sqh) [reply src/utils.py:13]: Aren't telegram bots also using public APIs? Currently, seems no other option to bypass the restrictions. Suggestions?

Comment (shailesh) [reply src/utils.py:13]: Any POC for telegram using public APIs?

Comment (shailesh) [reply src/utils.py:13]: let us not update the APIs

Comment (shivaslifzz) [reply src/utils.py:13]: it's not a good thing to add public api

**Comment by nolimits0007** [comment]
> fixes #448
> 
> Not tested. can someone test?

@6mxo89sqh can u plz give kvdb option in ur main code. It's the old kvdb key.

File: src/utils.py (hunk 1 above)
Comment (shivaslifzz) [reply src/utils.py:13]: for short term fix we can request the otp for every 3mins like we are doing for IOS

**Comment by shivaslifzz** [comment]
> > fixes #448
> > Not tested. can someone test?
> 
> @6mxo89sqh can u plz give kvdb option in ur main code. It's the old kvdb key.

use below command before  your python command 

export KVDB_BUCKET=URKEY

**Comment by nolimits0007** [comment]
> > > fixes #448
> > > Not tested. can someone test?
> > 
> > 
> > @6mxo89sqh can u plz give kvdb option in ur main code. It's the old kvdb key.
> 
> use below command before your python command
> 
> export KVDB_BUCKET=URKEY

Thanks

**Comment by nolimits0007** [comment]
Same error with new commits.

Screenshot 2021-06-03 at 20 57 48

**Comment by shivaslifzz** [comment]
I am able to book the slot using public API 

image

telegram notification at 9:05:13 and our code booked the slot at 09:05:21

image

**Comment by shivaslifzz** [comment]
> Same error with new commits.
> 
> Screenshot 2021-06-03 at 20 57 48

may be your using wrong branch please check out restrictions-on-refresh branch and use

**Comment by Nakul93** [comment]
I think for now Public API should be fine, I've tested it with 7sec freq and haven't got any rate limit issues.

**Comment by Nakul93** [comment]
> I am able to book the slot using public API
> 
> image
> 
> telegram notification at 9:05:13 and our code booked the slot at 09:05:21
> 
> image

Was this run on ec2?

**Comment by shivaslifzz** [comment]
> > I am able to book the slot using public API
> > image
> > telegram notification at 9:05:13 and our code booked the slot at 09:05:21
> > image
> 
> Was this run on ec2?

yes

**Comment by Nakul93** [comment]
Whats surprising is that telegram notification doesn't gets affected by these changes in the API, the alerts are always on time

**Comment by shivaslifzz** [comment]
> Whats surprising is that telegram notification doesn't gets affected by these changes in the API, the alerts are always on time

they are rerouting to the new IP when ever they reached to ip rate limit we can do this with aws but it's chargeable

**Comment by Nakul93** [comment]
I guess one workaround could be to use this telegram notification as a trigger to start the search as mentioned in #398. Only after getting the alert it should login and try to book.

**Comment by nolimits0007** [comment]
> > Same error with new commits.
> > Screenshot 2021-06-03 at 20 57 48
> 
> may be your using wrong branch please check out restrictions-on-refresh branch and use

Can u plz give the link for restrictions-on-refresh branch.

I am using this https://github.com/6mxo89sqh/covid-vaccine-booking.git

**Comment by Nakul93** [comment]
@nolimits0007 You need to checkout the pr, for this go into the main directory and type,
`git fetch origin pull/449/head:refresh` >>> this pulls the changes in this PR to a new branch called refresh
Now checkout the branch using , `git checkout refresh`

**Comment by nolimits0007** [comment]
Can anybody help with this error . I asked for Covaxine but script is booking Covieshield.

Screenshot 2021-06-03 at 21 51 51

**Comment by darthfishious** [comment]
> I think for now Public API should be fine, I've tested it with 7sec freq and haven't got any rate limit issues.

with public APIs there is no rate limit issue even with 1 second refresh rate. but caching is a problem. for me, it could not detect when a 1000 slot center opened up due to caching. It detected the center 5 minutes late, by which time it was all gone.

**Comment by shivaslifzz** [comment]
> > I think for now Public API should be fine, I've tested it with 7sec freq and haven't got any rate limit issues.
> 
> with public APIs there is no rate limit issue even with 1 second refresh rate. but caching is a problem. for me, it could not detect when a 1000 slot center opened up due to caching. It detected the center 5 minutes late, by which time it was all gone.

Just now I also faced same issue

**Comment by 6mxo89sqh** [comment]
> Can anybody help with this error . I asked for Covaxine but script is booking Covieshield.
> 
> Screenshot 2021-06-03 at 21 51 51

@nolimits0007 I am submitting a new PR which fixes this correctly.

**Comment by nolimits0007** [comment]
> > Can anybody help with this error . I asked for Covaxine but script is booking Covieshield.
> > Screenshot 2021-06-03 at 21 51 51
> 
> @nolimits0007 I am submitting a new PR which fixes this correctly.

@6mxo89sqh Thanks for addressing it.

**Comment by ggoel1998** [comment]
Can we bring back polling for every 1 second as we are using Public API

**Comment by ggoel1998** [comment]
> fixes #448
> 
> Not tested. can someone test?

already on it . working fine since last 20mins on aws ec2

**Comment by shailesh** [comment]
> Can we bring back polling for every 1 second as we are using Public API

We are using Protected ones

**Comment by mrms123** [comment]
Most likely, the rate limits are here to stay. So how about we do this
At the refresh frequency 'x' , we pull from public API
OTP expires 10 minutes, and we know we can do 20 refreshes in 10 minutes, i.e., one refresh every 30 seconds

But every nth refresh; n = ceil (30/x); we pull from protected API

We can parametrize this if needed, by taking OTP expire time as an input. And also maintain the refresh rate as value on some server (maybe even as a KVDB key) - since they have been playing around with this rate quite often.

For most users, they are going to be using defaults - which means as of now, they cannot refresh more than once in 30 secs without being rate limited. 
This just improves their chances. If there is no caching in public API - then they get immediately. If there is caching - then anyway once in 30 seconds we are checking using protected API. So the users are no worse off doing this, than they would have been, only using the protected API

Anyway @Nakul93 also mentions the public API has been working fine for him - this approach can help us get the best of both worlds

**Comment by Nakul93** [comment]
Yes I've tested it but not when slots are opening up. As many users have had problems where slots starts showing up late using Public API, it doesn't seem to be an efficient solution. 
Your suggestion is quite interesting but cowin constantly changes there requests limits.As of now its 20, it could be changed in the future and thereby changing the logic. 

> This just improves their chances. If there is no caching in public API - then they get immediately. If there is caching - then anyway once in 30 seconds we are checking using protected API. So the users are no worse off doing this, than they would have been, only using the protected API

30 secs is too much time. The slots will probably get booked by then

**Comment by mrms123** [comment]
> As of now its 20, it could be changed in the future and thereby changing the logic.

It doesn't have to. Same logic could apply. We could hold the request limit at some address, or a protected read only kvdb bucket (Can even just be a file on github, using a direct-link https://github.com/bombardier-gif/covid-vaccine-booking/raw/main/src/RATE.txt)

When limits on CoWIN is updated, this limit is updated. Force the code to check for this once every couple of hours (for increase in limit), or when rate limit is hit (for decrease in limit). This defines as variable - request_limit

then you can define the following:

polling_freq = 5 (or whatever user defines)

protected_freq = 600 / request_limit
elapsed_since_protected = protected_freq+1

if elapsed_since_protected >= protected_freq:
     check_calendar_via_protected
     elapsed_since_protected = polling_freq

else :
     check_calendar_via_public
     elapsed_since_protected += polling_freq


The fact of the matter remains - if we are rate limited - we are rate limited. 
I believe there is a dual check on IP + Token now. I tried rotating IPs with the same token - still ran into the same rate limit error. I tried changing token on same IP - still rate limited.

The other option, is to figure how to to bypass the CDN on the public API.

**Comment by Nakul93** [comment]
Consider a scenario in the above mentioned logic where the script is checking using the public APIs and a slot opens up but it shows up late due to caching making it unable to book.
Your findings seems to be solid, if you can submit a PR we can test it out.

**Comment by 6mxo89sqh** [comment]
Each rate limit change was different. Following have been the changes so far:
1. 100 request per IP
2. Forced rate limited after 10 minutes
3. 20 request per session.

So, I guess having a central / limit info will not help. Instead, I recommended to hard code this number in code itself. 

Your logic would work. (Using counter might be simple as u would not have to polling rate to each function) 

**Eagerly awaiting for your PR.**

**Comment by mrms123** [comment]
> Consider a scenario in the above mentioned logic where the script is checking using the public APIs and a slot opens up but it shows up late due to caching making it unable to book.

Agreed - we're going to miss out on that
let's think of it this way - if I had access to private API only - the best I could do is x refreshes per period given the limit. That x still holds, and we're checking at that frequency..

But in between - why not try our luck with the public API as well - in the alternate case, we would idle between the two refreshes.

I'll open a PR in a bit - I had added some other stuff of my own to the code, need to stash them.

**Pull request closed without merging**